### DISCUSSION TITLE
📦 Zipping casual loop as lookup or inter-dependent fraction from coflow structure

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#38] [people; relating]
Research hypothesis: When the desired fineness of time-scale cannot allow well-mixing of one compartment, conjugate function (defined as the same distributional family of prior and posterior) can be used for screen function.

Jose delved into exponential family for this, which I assume to be relevant to the fact conjugate prior can be designed for every exponential families "[For exponential families the likelihood is a simple standarized function of the parameter and we can define conjugate priors by mimicking the form of the likelihood.](https://people.eecs.berkeley.edu/~jordan/courses/260-spring10/other-readings/chapter9.pdf)".

The next step may be implementing exponential family on Hazhir's rework model. I uploaded initial points in conjugacy folder in conjugacy brach https://github.com/Data4DM/BayesSD/tree/conjugacy, https://github.com/Data4DM/BayesSD/tree/conjugacy/ContinuousCode/5_BayesCalib/experiment/conjugacy.

**Comment by hyunjimoon** [comment]
![image](https://user-images.githubusercontent.com/30194633/193871288-2e6812e5-edec-4a0c-b3e2-fde5e72aef8b.jpeg)

**Comment by tomfid** [comment]
I'm not following how conjugate priors help here. Is the screening function for detecting the problem, or something else?

In any case, I'm not sure I agree that fineness of time scale is the origin of the aggregation problem. Consider the following thought experiment.

Red and blue people go to a bar. They arrive at a uniform rate, or stationary random rate if you prefer. Once in the room, they mingle with others. After an exponentially-distributed time they leave. This is basically the classic first-order well mixed system. Suppose we're interested in the average wealth in the bar, and reds have twice as much money as blues. Then we could measure the flow of colors on the way in the door and infer the average stock from that.

This can go wrong in several ways.

1. The residence time process is not really exponential. Perhaps people enter at the N door, work their way across the room, and leave at the S door, with a fixed transit time, or some higher-order distribution (Erlang(10)).
2. The residence time is not uniform by class. Blues stay twice as long as reds.
3. There are nonlinear interactions within the color groups. For example, if one color group perceives dominance above some threshold, it initiates a bar fight and quickly ejects the other color group.

#1 implies that the first-order structure is wrong both in aggregate and for any individual agent. You need an aging chain or pipeline delay, or maybe a spatial grid, to represent the process. #2 implies that you need to disaggregate by color (but you can keep the first-order assumption). #3 probably requires disaggregation by color, and some additional feedback among the colors that varies the residence time according to the fight behavior.

I'm not sure any of these is directly related to the time scale, except in the sense that some of these behaviors may be too fine-scaled for us to care about if the model horizon is long. For example, in case #3, if we're modeling the bar's financial health over 10 years, we don't care about the high frequency behavior of bar fights and color distinctions. We could just roll that into a slightly shorter aggregate time constant. But if we're modeling a night, or even a month, we probably do care about these details.

When I hear "fineness of time scale cannot allow well mixing" I think of a literal mixing problem. The bar is full of blue people. 10 reds enter through the door. After a couple of minutes, you have reds and blues mingled. But one second after the reds enter, they're probably still in a clump by the door. Does this matter? Maybe - it depends again on the time horizon of interest for the model, and whether the temporary spatial clumping initiates any nonlinear processes that are important enough to require disaggregation.

**Comment by hyunjimoon** [comment]
Design convention enforcing temporal cauchy convergence w.r.t. time steps (halve ts until stock values(ts) ~ stock values(ts/2)) but not space step (i.e. subscript) is the source of the problem.

May I summarize your setting as: measuring quantities red and blue differs (f(theta1, theta2,t); not f(theta,t))? Two comments:

1. f(theta1, thera2,t) derails from poisson process with exponential delay and becomes compound and/or conditional poisson process. The former is infinitely divisible having stationary and independent increments whereas the latter has only stationary increments (not independent). Conditional PP cannot have independent increments as its rate lambda is random variable unlike compound process (its lambda is scalar). Considering scalar is projected random variable (~ using base point instead of lookup function), seemingly different outcomes of the two process is once again the _artifact_ of our choice especially time step; if our chosen timestep is small enough to have constant rate, then our outcome is compound poisson process. If the rate is homogeneous across time steps but heterogeneous within one timestep, it becomes conditional poisson process. If the rate is heterogeneous across timesteps but homogeneous within one time step, this may (I don't know any analytical process which is why we run simulation) be non homogeneous compound poisson process.

@tomfid I wish to at least get your consent on random variable ~ lookup function recalling our distribution function discussion last week.

This needs justification but 'Time scale' seems to be the finest lego block we build for calculus system (Lebesgue integral). "fine enough time step?" is the other side of "Long enough time horizon?" question as we want compact (closed and bounded) integral interval and once it is compact we can scale it without losing any info.

2. linger time and wealth are known to be different among red and blue. But the starting point is again wrong. It does not start from the question of "are red and blue the same?" but "given a time series of 100 people, what is the best cluster i.e. good enough well-mixed subgroup"

red and blue ppl: "After an exponentially-distributed time they leave. This is basically the classic first-order well mixed system". As long as we measure

**Comment by tomfid** [comment]
1. I don't think random distributions matter to the aggregation question. The red/blue problem is the same in a continuous/deterministic system with an idealized input like a step function.
2. Not sure what the proposal is here. "Are red and blue the same" seems like the same question as "is clustering red and blue good enough", at least in a practical sense. Or do you mean something more complex, like labeling people on the way in the door according to a vector of attributes for which we don't necessarily know the effects, then trying to infer which ones matter?

I think the point about easy enforcement of temporal convergence is insightful. I think it's a consequence of the language architecture - time can be changed globally with a switch, but space, aging chain order, red/blueness, etc. require manual labor to change subscripts typically. It would be interesting to think through a simple way of expressing something more flexible. Attribute coflows are also something that could be automated, but isn't.

**Reply by hyunjimoon** [reply]
> easy enforcement of temporal convergence is insightful. 
Thanks for the paraphrase. Another paraphrase is done at the head of https://github.com/Data4DM/BayesSD/issues/17 i.e. Bayequentist chain management goal as "online decision, dynamic timestep aggregation, dynamic subscript aggregation".

**Comment by hyunjimoon** [comment]
SD's logistic choice forumation is similar to dirichlet process's prior 
X = X1 + X2 (heterogeneous subgroup)
$\frac{dX}{dt} = -c_0 X$ 
i) $\frac{dX_1}{dt} = - c_1 X_1$
ii) $\frac{dX_2}{dt} = - c_2 X_2$
$X_1 =  \frac{e^{c_1}}{(e^{c_1} +  e^{c_2})} X, X_2 =  \frac{e^{c_2}}{(e^{c_1} +  e^{c_2})} X$

## SD
<img width="1089" alt="image" src="https://user-images.githubusercontent.com/30194633/194543773-2e403ab3-bca9-4a28-b7f5-bf66d2655bff.png">

## Bayes
theta_j = exp(phi_j)/(exp(phi_1)+exp(phi_2)+exp(phi_3)+exp(phi_4) suggested as answer to "priors to dirichlet prior" in [Gelblog](https://statmodeling.stat.columbia.edu/2009/04/29/conjugate_prior/) and equation 1 from his [paper](http://www.stat.columbia.edu/~gelman/research/published/bois2.pdf) "Pharmacokinetic Analysis using population modeling and informative prior distributions". Sec.2.3 has great info on prior dist.
- can we view the lower coflow loop as a whole as one lookup mapping layer_n parameter with layer_{n+1} parameter?
-  Could allocation be used for heterogenous outflow in coflow? Why does logit function appear from both lower left of James and Jeorun's poster?


[20220710_Sust_Consumption_ISDC 2022_Poster_v02.pdf](https://github.com/Data4DM/BayesSD/files/9732939/20220710_Sust_Consumption_ISDC.2022_Poster_v02.pdf)
[SJDM Poster - Modeling Dynamic Supply Chains with Endogenous Allocations and Market Clearing.pdf](https://github.com/Data4DM/BayesSD/files/9737939/SJDM.Poster.-.Modeling.Dynamic.Supply.Chains.with.Endogenous.Allocations.and.Market.Clearing.pdf)

**Reply by hyunjimoon** [reply]
Coflow examples in building renewal, labor force and employee and thier options
![image](https://user-images.githubusercontent.com/30194633/195131059-41c35911-1d40-412e-b2ad-1b29308997f0.png)

![D9421BD3-4168-4CF5-BB77-7E6386484ED9_1_201_a](https://user-images.githubusercontent.com/30194633/194541896-45bab586-96f5-4ac9-aaed-f9eafd1001ea.jpeg)

**Comment by hyunjimoon** [comment]
For some reason Hazhir updated the coflow units (from left to right) and I wonder whether the right version is more communicable. 

![image](https://user-images.githubusercontent.com/30194633/203606220-d729075e-ce0a-4812-bc9f-620a54acd204.png)

first version has different unit for coflow (task for the above, defect for below), but recent Hazhir's version improved(?) this by using the same unit, naming the conditioned coflow `Defects ~` as `Defective Tasks ~`. Traditional coflow is conditioning (propagating uncertainty with sequential generation) as its dependence structure is one-sided. This is in contrast to prey-predator where two stock-chains affects each other. Hazhir's rework doesn't seem to be traditional coflow, though as the output of Defective-task flow affects the Task flow again via `Ave Dfct in Tstd Task (zidz of task to be tested and defective task to be tested)` - `Test Failure Frac` - `rejecting task`. (need more reflection, feedback welcome!)

**Reply by tomfid** [reply]
Interesting. I think this is more than a change of units. The left is a coflow. The right is disaggregation. The one on the right feels more natural to me, but might be less useful when errors are nonbinary.

**Reply by hyunjimoon** [reply]
I changed the original title "add conjugacy for distributional representation (e.g. coflows into theta -> theta|x)" to "interlocked coflow fraction". Imagine a hierarchy with variable $\tau$ closer to the root than $\theta$. Update of $\theta_{t}|\tau_t \rightarrow \theta_{t+1}|\tau_t$ and $\tau_{t}|\theta_{t+1} \rightarrow \tau_{t+1}|\theta_{t+1}$ seems to be in action.

**Reply by hyunjimoon** [reply]
Do you think every stock should have a physical correspondence? What are the goals of the model (i.e. which quantities of interest to track) that can make coflow and aggregation formulation similar?

I think this is tangentially relevant to your comment on physical existence https://github.com/Data4DM/BayesSD/discussions/94#discussioncomment-4170636

**Reply by tomfid** [reply]
I think every stock should correspond with something we think exists (though there are some convenience uses I might excuse), but that doesn't mean it's necessarily physical.

**Reply by tomfid** [reply]
This is a pretty interesting question. Consider a vehicle fleet model.

Coflow: you have stocks of vehicles, and fuel consumption, and can compute the average fuel economy.
Disaggregate: you have stocks of cars (high fuel economy) and trucks (low fuel economy) and can compute the average for the mix.
Entity: you keep track of vehicles in every model-year combination.

The coflow formulation can't keep track of the mix of vehicles on the road. You also can't handle things like different lifetimes for cars & trucks, or interactions among types. However, it's easy to add more attributes (weight) as additional coflows to the master stock.

The disaggregate formulation can't handle changing fuel economy within the types (unless you add a coflow to each!). Also, if you want more attributes, you might get into a combinatorial explosion: (heavy, light) x (efficient, inefficient) x (car, truck).

At some point you need the entity formulation, because the actual model-year options are less numerous than the combinatorial disaggregation, and easier to manage than a lot of coflows.

In real settings I recall, it was kind of rare to really need the entity version - usually some kind of hybrid disaggregate-coflow could handle anything truly needed - but sometimes the entity version is easier to construct and more transparent, because you don't have to ponder aggregation rules.

**Comment by hyunjimoon** [comment]
![image](https://user-images.githubusercontent.com/30194633/205794431-7c2c2238-a5d9-4421-9dbc-2159857335c1.png)

**Reply by hyunjimoon** [reply]
Starting from #108, Tom shared the above model with comments:
>  it's not right. I think I made the classic error in us/(us+them) allocation that I was trying to describe, which would be great except that I was trying not to make the error
> constrains the allocation a little better, and adds perceptions of errors and delays

The most interesting feature is time stock. @tomfid Have you used this technique in other model before? Relevant to our discussion on dynamic temporal-spatial aggregation and time step #34, I wonder how time stock would look when translated into stanify language. 

- dy/dt computed for every stock variable: unit change of time stock?
- [cumulative time (save_prs) function](https://github.com/Data4DM/stanify/commit/e4478e434a1cd545645030a8d01a92cdf1bb8d4b#diff-2b0e808f1f1a921e4434f1d86945af9efc4b4885bc5c6e05648d17eea07c8350)

**Comment by hyunjimoon** [comment]
I think [this book](https://www.amazon.com//dp/B0BG8ZY8MG?pd_rd_i=B0BG8ZY8MG&pd_rd_w=aB5G1&content-id=amzn1.sym.ac1fafc2-1bb1-4581-9a9f-b42ca4d9089a&pf_rd_p=ac1fafc2-1bb1-4581-9a9f-b42ca4d9089a&pd_rd_wg=yJOTX&pd_rd_r=d5a9d1b7-30dd-46eb-a5a0-eb16ec013ac6&ref_=pe_42015120_689071590) published today would be relevant to our conjugate prior design:

During the past half-century, exponential families have attained a position at the center of parametric statistical inference. Theoretical advances have been matched, and more than matched, in the world of applications, where logistic regression by itself has become the go-to methodology in medical statistics, computer-based prediction algorithms, and the social sciences. This book is based on a one-semester graduate course for first year Ph.D. and advanced master's students. After presenting the basic structure of univariate and multivariate exponential families, their application to generalized linear models including logistic and Poisson regression is described in detail, emphasizing geometrical ideas, computational practice, and the analogy with ordinary linear regression. Connections are made with a variety of current statistical methodologies: missing data, survival analysis and proportional hazards, false discovery rates, bootstrapping, and empirical Bayes analysis. The book connects exponential family theory with its applications in a way that doesn't require advanced mathematical preparation.
